### PR TITLE
Implement Block Reward Reallocation

### PR DESCRIPTION
**Pull request by UdjinM6**

### CODE REVIEW

File: src/chainparams.cpp [hunk 1]
```diff
@@ -342,6 +342,13 @@ class CMainParams : public CChainParams {
         consensus.vDeployments[Consensus::DEPLOYMENT_DIP0008].nWindowSize = 4032;
         consensus.vDeployments[Consensus::DEPLOYMENT_DIP0008].nThreshold = 3226; // 80% of 4032
 
+        // Deployment of Block Reward Reallocation
+        consensus.vDeployments[Consensus::DEPLOYMENT_REALLOC].bit = 5;
+        consensus.vDeployments[Consensus::DEPLOYMENT_REALLOC].nStartTime = 1598918400; // Sep 1st, 2020
+        consensus.vDeployments[Consensus::DEPLOYMENT_REALLOC].nTimeout = 1630454400; // Sep 1st, 2021
```
Comment (PastaPastaPasta) [inline src/chainparams.cpp:348]: Change this to October 1st?

File: src/masternode/masternode-payments.cpp [hunk 2]
```diff
@@ -313,13 +313,22 @@ bool CMasternodePayments::GetBlockTxOuts(int nBlockHeight, CAmount blockReward,
...
+
+        const Consensus::Params& consensusParams = Params().GetConsensus();
+        if (VersionBitsState(pindex, consensusParams, Consensus::DEPLOYMENT_REALLOC, versionbitscache) == ThresholdState::ACTIVE) {
+            nReallocActivationHeight = VersionBitsStateSinceHeight(pindex, consensusParams, Consensus::DEPLOYMENT_REALLOC, versionbitscache);
+        }
     }
+
+    CAmount masternodeReward = GetMasternodePayment(nBlockHeight, blockReward, nReallocActivationHeight);
+
+    uint256 proTxHash;
```
Comment (PastaPastaPasta) [inline src/masternode/masternode-payments.cpp:331]: What's the point of this?

**Review by PastaPastaPasta — CHANGES_REQUESTED** [review]
couple of changes

File: test/functional/feature_block_reward_reallocation.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+Checks block reward reallocation correctness
+
+'''
+
+class BlockRewardReallocationTest(DashTestFramework):
+    def set_test_params(self):
+        self.set_dash_test_params(4, 3, fast_dip3_enforcement=True)
+        self.set_dash_dip8_activation(450)
+
+    def create_test_block(self, version=536870912):
```
Comment (PastaPastaPasta) [inline test/functional/feature_block_reward_reallocation.py:23]: what's up with this magic number?

Comment (UdjinM6) [reply test/functional/feature_block_reward_reallocation.py:23]: it's `0x20000000` i.e. not signalling for anything

File: src/chainparams.cpp (hunk 1 above)
Comment (UdjinM6) [reply src/chainparams.cpp:348]: This date doesn't have to be in the future really, just smth close to the release, but ok.

File: test/functional/feature_block_reward_reallocation.py (hunk 3 above)
Comment (PastaPastaPasta) [reply test/functional/feature_block_reward_reallocation.py:23]: Ahh okay, should we add that info as a comment? like 
```suggestion
    # 536870912 == 0x20000000, i.e. not signalling for anything
    def create_test_block(self, version=536870912):
```

File: src/masternode/masternode-payments.cpp (hunk 2 above)
Comment (UdjinM6) [reply src/masternode/masternode-payments.cpp]: 🙈

**Comment by xdustinface** [comment]
Can we add af80d7514816b343225068f2c6752ce0805d1c75 + cb0c76ffd666fa06c36c57fc905c35f9617d8916 to refactor the calculations.

And the new test file isn't executable 9e739705e0cdab413b5864c5a122a5fd3760609f

**Review by xdustinface — APPROVED** [review]
Looks good, nice tests, they seem to test what they should and they pass, (ut)ACK

**Review by PastaPastaPasta — APPROVED** [review]
ACK, seems to work as expected. Mainnet syncs

**Pull request merged**